<center>
    <img src="https://gitlab.com/ibm/skills-network/courses/placeholder101/-/raw/master/labs/module%201/images/IDSNlogo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>


# **Launch Sites Locations Analysis with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [1]:
!pip3 install folium
!pip3 install wget

In [2]:
import folium
import wget
import pandas as pd

In [3]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module\_3/DV0101EN-3-5-1-Generating-Maps-in-Python-py-v2.0.ipynb)


## Task 1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [4]:
# Download and read the `spacex_launch_geo.csv`
spacex_csv_file = wget.download('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv')
spacex_df=pd.read_csv(spacex_csv_file)

Now, you can take a look at what are the coordinates for each site.


In [5]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610746

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [6]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [7]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#0000FF', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small blue circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*TODO:*  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [8]:
# Initialize the map
site_map = folium.Map(location=nasa_coordinate, zoom_start=4)
# For each launch site, add a Circle object based on its coordinate (Lat, Long) values. In addition, add Launch site name as a popup label
for i in range(launch_sites_df.shape[0]):
    site=launch_sites_df.iloc[i][0]
    #print(site)
    lat =launch_sites_df.iloc[i][1]
    #print(lat)
    long = launch_sites_df.iloc[i][2]
    #print(long)
    circle = folium.Circle([lat,long], radius=1000, color='#0000FF', fill=True).add_child(folium.Popup(str(site)))
    marker = folium.map.Marker(
        [lat,long],
        # Create an icon as a text label
        icon=DivIcon(
            icon_size=(20,20),
            icon_anchor=(0,0),
            html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % str(site),
            )
        )
    site_map.add_child(circle)
    site_map.add_child(marker)

site_map.add_child(circle)
site_map.add_child(marker)

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png" />
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?
*   Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


# Task 2: Mark the success/failed launches for each site on the map


Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [9]:
spacex_df.tail(10)

Launch Site        Lat       Long  class
46    KSC LC-39A  28.573255 -80.646895      1
47    KSC LC-39A  28.573255 -80.646895      1
48    KSC LC-39A  28.573255 -80.646895      1
49  CCAFS SLC-40  28.563197 -80.576820      1
50  CCAFS SLC-40  28.563197 -80.576820      1
51  CCAFS SLC-40  28.563197 -80.576820      0
52  CCAFS SLC-40  28.563197 -80.576820      0
53  CCAFS SLC-40  28.563197 -80.576820      0
54  CCAFS SLC-40  28.563197 -80.576820      1
55  CCAFS SLC-40  28.563197 -80.576820      0

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


*TODO:* Create a new column in `launch_sites` dataframe called `marker_color` to store the marker colors based on the `class` value


In [10]:
# Function to assign color to launch outcome
def assign_marker_color(launch_outcome):
    if launch_outcome == 1:
        return 'green'
    else:
        return 'red'
    
spacex_df['marker_color'] = spacex_df['class'].apply(assign_marker_color)
spacex_df.tail(10)

Launch Site        Lat       Long  class marker_color
46    KSC LC-39A  28.573255 -80.646895      1        green
47    KSC LC-39A  28.573255 -80.646895      1        green
48    KSC LC-39A  28.573255 -80.646895      1        green
49  CCAFS SLC-40  28.563197 -80.576820      1        green
50  CCAFS SLC-40  28.563197 -80.576820      1        green
51  CCAFS SLC-40  28.563197 -80.576820      0          red
52  CCAFS SLC-40  28.563197 -80.576820      0          red
53  CCAFS SLC-40  28.563197 -80.576820      0          red
54  CCAFS SLC-40  28.563197 -80.576820      1        green
55  CCAFS SLC-40  28.563197 -80.576820      0          red

*TODO:* For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [11]:
# instantiate a mark cluster object for the launches in the dataframe
launches = MarkerCluster().add_to(site_map)

# loop through the dataframe and add each data point to the mark cluster
for lat, lng, cls, clr in zip(spacex_df.Lat, spacex_df.Long, spacex_df['class'], spacex_df.marker_color):
    if cls==1:
        label='Successful Launch!'
    else:
        label='Failed Launch'
    folium.Marker(
        location=[lat, lng],
        popup=label,
        icon=folium.Icon(color=str(clr))
    ).add_to(launches)

# display map
site_map

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png" />
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png" />
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


# TASK 3: Calculate the distances between a launch site to its proximities


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [12]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


You can calculate the distance between two points on the map based on their `Lat` and `Long` values using the following method:


In [13]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

*TODO:* Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [14]:
# find coordinate of the closet coastline
# e.g.,: Lat: 28.56367  Lon: -80.57163
coastline1 = [28.56399,-80.56811]
distance_coastline = round(calculate_distance(launch_sites_df.iloc[1][1], launch_sites_df.iloc[1][2], coastline1[0], coastline1[1]),3)
print('Coastline is %s kilometers away.'%distance_coastline)

Coastline is 0.855 kilometers away.


*TODO:* After obtained its coordinate, create a `folium.Marker` to show the distance


In [15]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
# for example
distance_marker = folium.Marker(
coastline1,
icon=DivIcon(
    icon_size=(20,20),
    icon_anchor=(0,0),
    html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_coastline),
    )
)

*TODO:* Draw a `PolyLine` between a launch site to the selected coastline point


In [16]:
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
lines=folium.PolyLine(locations=[nasa_coordinate,coastline1], weight=1,color='blue')

Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png" />
</center>


*TODO:* Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png" />
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png" />
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png" />
</center>


### Adding coordinates of features near the launch sites

In [17]:
# Coordinates of features near the launch sites:
nearby = {'Launch Site':['CCAFS LC-40','CCAFS SLC-40','KSC LC-39A','VAFB SLC-4E'],'Rail Lat':[28.57207,28.57207,28.57291,34.6337],'Rail Long':[-80.58528,-80.58528,-80.65396,-120.62473],'Highway Lat':[28.56376,28.56376,28.57327,34.67],'Highway Long':[-80.57084,-80.57084,-80.65547,-120.55919], 'Ocean Lat':[28.56285,28.56285,28.57518,34.63478], 'Ocean Long':[-80.56792,-80.56792,-80.57136,-120.6253],'Nearest City Lat':[28.61353,28.61353,28.61263,34.63872],'Nearest City Long':[-80.80574,-80.80574,-80.80574,-120.45752]}
# Merging the feature coordinates with the launch site coordinates:
nearby_df= pd.DataFrame(nearby)
lat_lon = launch_sites_df.merge(nearby_df, on='Launch Site')
lat_lon

Launch Site        Lat        Long  Rail Lat  Rail Long  Highway Lat  \
0   CCAFS LC-40  28.562302  -80.577356  28.57207  -80.58528     28.56376   
1  CCAFS SLC-40  28.563197  -80.576820  28.57207  -80.58528     28.56376   
2    KSC LC-39A  28.573255  -80.646895  28.57291  -80.65396     28.57327   
3   VAFB SLC-4E  34.632834 -120.610746  34.63370 -120.62473     34.67000   

   Highway Long  Ocean Lat  Ocean Long  Nearest City Lat  Nearest City Long  
0     -80.57084   28.56285   -80.56792          28.61353          -80.80574  
1     -80.57084   28.56285   -80.56792          28.61353          -80.80574  
2     -80.65547   28.57518   -80.57136          28.61263          -80.80574  
3    -120.55919   34.63478  -120.62530          34.63872         -120.45752

### Calculating the distance between the features and launch sites
### Adding the distance to each feature
### Adding the lines between features and launch sites

In [18]:
Rail_dist = []
Highway_dist = []
Ocean_dist = []
City_dist = []

for i, row in lat_lon.iterrows():
    # Launch Site Coords:
    lat1 = lat_lon.iloc[i][1]
    lon1 = lat_lon.iloc[i][2]
    # Rail Coords and dist:
    latR = lat_lon.iloc[i][3]
    lonR = lat_lon.iloc[i][4]
    Rail_dist.append(calculate_distance(lat1, lon1, latR, lonR))
    Rail_lines=folium.PolyLine(locations=[[lat1, lon1],[latR,lonR]], weight=1,color='yellow')
    site_map.add_child(Rail_lines)
    if i!=1:
        rail_dist_marker = folium.Marker([latR,lonR],icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0),html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(Rail_dist[i])))
        site_map.add_child(rail_dist_marker)
    # Highway Coords and dist:
    latH = lat_lon.iloc[i][5]
    lonH = lat_lon.iloc[i][6]
    Highway_dist.append(calculate_distance(lat1, lon1, latH, lonH))
    highway_lines=folium.PolyLine(locations=[[lat1, lon1],[latH,lonH]], weight=1,color='green')
    site_map.add_child(highway_lines)
    if i!=1:
        highway_dist_marker = folium.Marker([latH,lonH],icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0),html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(Highway_dist[i])))
        site_map.add_child(highway_dist_marker)
    # Ocean Coords and dist:
    latO = lat_lon.iloc[i][7]
    lonO = lat_lon.iloc[i][8]
    Ocean_dist.append(calculate_distance(lat1, lon1, latO, lonO))
    ocean_lines=folium.PolyLine(locations=[[lat1, lon1],[latO,lonO]], weight=1,color='blue')
    site_map.add_child(ocean_lines)
    if i!=1:
        ocean_dist_marker = folium.Marker([latO,lonO],icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0),html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(Ocean_dist[i])))
        site_map.add_child(ocean_dist_marker)
    # City Coords and dist:
    latC = lat_lon.iloc[i][9]
    lonC = lat_lon.iloc[i][10]
    City_dist.append(calculate_distance(lat1, lon1, latC, lonC))
    city_lines=folium.PolyLine(locations=[[lat1, lon1],[latC,lonC]], weight=1,color='red')
    site_map.add_child(city_lines)
    if i!=1:
        city_dist_marker = folium.Marker([latC,lonC],icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0),html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(City_dist[i])))
        site_map.add_child(city_dist_marker)

# Final Map

In [19]:
site_map

### Creating one dataframe with all of our calculations

In [20]:
# Create a marker with distance to a closest city, railway, highway, etc.
dist = {'Launch Site':['CCAFS LC-40','CCAFS SLC-40','KSC LC-39A','VAFB SLC-4E'],'Rail_dist [km]':Rail_dist,'Highway_dist [km]':Highway_dist,'Ocean_dist [km]':Ocean_dist,'City_dist [km]':City_dist}
dist_df = pd.DataFrame(dist)
full_df = lat_lon.merge(dist_df, on='Launch Site')
full_df

Launch Site        Lat        Long  Rail Lat  Rail Long  Highway Lat  \
0   CCAFS LC-40  28.562302  -80.577356  28.57207  -80.58528     28.56376   
1  CCAFS SLC-40  28.563197  -80.576820  28.57207  -80.58528     28.56376   
2    KSC LC-39A  28.573255  -80.646895  28.57291  -80.65396     28.57327   
3   VAFB SLC-4E  34.632834 -120.610746  34.63370 -120.62473     34.67000   

   Highway Long  Ocean Lat  Ocean Long  Nearest City Lat  Nearest City Long  \
0     -80.57084   28.56285   -80.56792          28.61353          -80.80574   
1     -80.57084   28.56285   -80.56792          28.61353          -80.80574   
2     -80.65547   28.57518   -80.57136          28.61263          -80.80574   
3    -120.55919   34.63478  -120.62530          34.63872         -120.45752   

   Rail_dist [km]  Highway_dist [km]  Ocean_dist [km]  City_dist [km]  
0        1.334019           0.656942         0.923887       23.022304  
1        1.287251           0.587545         0.870321       23.048590  
2        0.691165           0.837603         7.381534       16.119865  
3        1.283486           6.272404         1.349495       14.038146

After you plot distance lines to the proximities, you can answer the following questions easily:

*   Are launch sites in close proximity to railways?
*   Are launch sites in close proximity to highways?
*   Are launch sites in close proximity to coastline?
*   Do launch sites keep certain distance away from cities?

Also please try to explain your findings.


#### Answers to the above questions:
* Yes, all launch sites are within 1.4 km of a railway. This makes sense as many of the heavier components will arrive as the launch site by rail.
* Yes, for the most part. Three of the four launch sites are within 0.9 km of a highway and the last site is 6.3 km from a highway. This also makes sense as there will be many components arriving by truck.
* Yes, for the most part. Three of the four launch sites are within 1.4 km of a coastline and the last site is 7.4 km from a coastline.  
* Yes, all launch sites are at least 14 km from a city. The proximity to coastlines and separation from cities can both be attributed to the safety of nearby populations. It would be preferential for the debris from a failed launch to end up in the ocean than over a highly populated area. 

# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Yan Luo](https://www.linkedin.com/in/yan-luo-96288783/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2021-01-01)


### Other Contributors


Joseph Santarcangelo


## Change Log


| Date (YYYY-MM-DD) | Version | Changed By | Change Description          |
| ----------------- | ------- | ---------- | --------------------------- |
| 2021-05-26        | 1.0     | Yan        | Created the initial version |


Copyright © 2021 IBM Corporation. All rights reserved.
